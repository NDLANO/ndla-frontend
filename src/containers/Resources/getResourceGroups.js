import defined from 'defined';
import {
  RESOURCE_TYPE_LEARNING_PATH,
  RESOURCE_TYPE_SUBJECT_MATERIAL,
  RESOURCE_TYPE_TASKS_AND_ACTIVITIES,
  RESOURCE_TYPE_ASSESSMENT_RESOURCES,
  RESOURCE_TYPE_SOURCE_MATERIAL,
  RESOURCE_TYPE_EXTERNAL_LEARNING_RESOURCES,
} from '../../constants';

export const sortOrder = {
  [RESOURCE_TYPE_LEARNING_PATH]: 1,
  [RESOURCE_TYPE_SUBJECT_MATERIAL]: 2,
  [RESOURCE_TYPE_TASKS_AND_ACTIVITIES]: 3,
  [RESOURCE_TYPE_ASSESSMENT_RESOURCES]: 4,
  [RESOURCE_TYPE_SOURCE_MATERIAL]: 5,
  [RESOURCE_TYPE_EXTERNAL_LEARNING_RESOURCES]: 6,
};

export const groupResourcesByResourceTypes = (
  supplementaryResources,
  coreResources,
) => {
  const resources = [
    ...coreResources,
    ...supplementaryResources
      .map(resource => ({
        ...resource,
        additional: true,
      }))
      .filter(resource => !coreResources.find(core => core.id === resource.id)), // don't show supp resources that exists in core
  ];
  return resources.reduce((obj, resource) => {
    const resourceTypesWithResources = resource.resourceTypes?.map(type => {
      const existing = defined(obj[type.id], []);
      return { ...type, resources: [...existing, resource] };
    });
    const reduced = resourceTypesWithResources?.reduce(
      (acc, type) => ({ ...acc, [type.id]: type.resources }),
      {},
    );
    return { ...obj, ...reduced };
  }, {});
};

export const sortResourceTypes = resourceTypes =>
  [...resourceTypes].sort((a, b) => {
    if (sortOrder[a.id] === undefined && sortOrder[b.id] === undefined)
      return 0;
    if (sortOrder[a.id] === undefined) return 1;
    if (sortOrder[b.id] === undefined) return -1;
    if (sortOrder[a.id] > sortOrder[b.id]) return 1;
    if (sortOrder[a.id] < sortOrder[b.id]) return -1;
    return 0;
  });

export const getResourceGroups = (
  resourceTypes,
  supplementaryResources,
  coreResouces,
) => {
  const groupedResources = groupResourcesByResourceTypes(
    supplementaryResources,
    coreResouces,
  );
  const sortedResourceTypes = sortResourceTypes(resourceTypes);

  return sortedResourceTypes
    .map(type => {
      const resources = defined(groupedResources[type.id], []);
      return { ...type, resources };
    })
    .filter(type => type.resources.length > 0);
};
