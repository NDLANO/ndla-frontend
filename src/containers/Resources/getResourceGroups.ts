/**
 * Copyright (c) 2018-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { sortBy, uniqBy } from "@ndla/util";
import {
  RESOURCE_TYPE_LEARNING_PATH,
  RESOURCE_TYPE_SUBJECT_MATERIAL,
  RESOURCE_TYPE_TASKS_AND_ACTIVITIES,
  RESOURCE_TYPE_ASSESSMENT_RESOURCES,
  RESOURCE_TYPE_SOURCE_MATERIAL,
  RESOURCE_TYPE_CONCEPT,
} from "../../constants";
import { GQLResource, GQLResourceType } from "../../graphqlTypes";
import { contentTypeMapping } from "../../util/getContentType";

export const sortOrder: Record<string, number> = {
  [RESOURCE_TYPE_LEARNING_PATH]: 1,
  [RESOURCE_TYPE_SUBJECT_MATERIAL]: 2,
  [RESOURCE_TYPE_TASKS_AND_ACTIVITIES]: 3,
  [RESOURCE_TYPE_ASSESSMENT_RESOURCES]: 4,
  [RESOURCE_TYPE_SOURCE_MATERIAL]: 5,
  [RESOURCE_TYPE_CONCEPT]: 6,
  default: 7,
};

type GQLResourceLike = Pick<GQLResource, "id" | "resourceTypes" | "rank" | "relevanceId">;

export const sortResources = <T extends GQLResourceLike>(resources: T[], isGrouped?: boolean) => {
  const uniq = uniqBy(resources, (res) => res.id);
  const sortedByRank = sortBy(uniq, (res) => res.rank ?? res.id);
  if (!isGrouped) {
    // TODO: Just return sortedByRank once we can remove contentType
    return sortedByRank.map((res) => {
      const firstResourceType = sortResourceTypes(res.resourceTypes ?? [])?.[0];
      return { ...res, contentType: firstResourceType ? contentTypeMapping[firstResourceType.id] : undefined };
    });
  }

  // TODO: Convert map to `sortBy` once we can remove contentType
  const withContentType = sortedByRank.map((res) => {
    const firstResourceType = sortResourceTypes(res.resourceTypes ?? [])?.[0];
    return { ...res, contentType: contentTypeMapping[firstResourceType?.id ?? "default"] };
  });
  return sortBy(withContentType, (res) => res.contentType);
};

type SharedResourceType = Pick<GQLResourceType, "id" | "name">;

export const sortResourceTypes = (resourceTypes: SharedResourceType[]) =>
  sortBy(resourceTypes, (type) => sortOrder[type.id] ?? sortOrder.default);
