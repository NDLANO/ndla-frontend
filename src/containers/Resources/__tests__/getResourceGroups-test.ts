/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { resourceTypes } from "./mockResources";
import { sortResourceTypes } from "../getResourceGroups";

test("resources types sort order", () => {
  const types = sortResourceTypes(resourceTypes);

  expect(types.length).toBe(6);
  expect(types[0]?.id).toBe("urn:resourcetype:learningPath");
  expect(types[1]?.id).toBe("urn:resourcetype:subjectMaterial");
  expect(types[2]?.id).toBe("urn:resourcetype:tasksAndActivities");
  expect(types[3]?.id).toBe("urn:resourcetype:reviewResource");
  expect(types[4]?.id).toBe("urn:resourcetype:SourceMaterial");
  expect(types[5]?.id).toBe("urn:resourcetype:concept");

  expect(types).toMatchSnapshot();
});
