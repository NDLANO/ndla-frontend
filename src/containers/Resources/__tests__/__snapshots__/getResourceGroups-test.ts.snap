// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resources types sort order 1`] = `
[
  {
    "id": "urn:resourcetype:learningPath",
    "name": "Læringssti",
  },
  {
    "id": "urn:resourcetype:subjectMaterial",
    "name": "Fagstoff",
    "subtypes": [
      {
        "id": "urn:resourcetype:academicArticle",
        "name": "Fagartikkel",
      },
      {
        "id": "urn:resourcetype:movieAndClip",
        "name": "Film og filmklipp",
      },
      {
        "id": "urn:resourcetype:simulation",
        "name": "Simulering",
      },
    ],
  },
  {
    "id": "urn:resourcetype:tasksAndActivities",
    "name": "Oppgaver og aktiviteter",
    "subtypes": [
      {
        "id": "urn:resourcetype:exercise",
        "name": "Øvelse",
      },
      {
        "id": "urn:resourcetype:task",
        "name": "Oppgave",
      },
      {
        "id": "urn:resourcetype:workAssignment",
        "name": "Arbeidsoppdrag",
      },
    ],
  },
  {
    "id": "urn:resourcetype:reviewResource",
    "name": "Vurderingsressurs",
  },
  {
    "id": "urn:resourcetype:SourceMaterial",
    "name": "Kildemateriale",
  },
  {
    "id": "urn:resourcetype:concept",
    "name": "Forklaring",
    "subtypes": [
      {
        "id": "urn:resourcetype:conceptArticle",
        "name": "Forklaringsartikkel",
      },
    ],
  },
]
`;
